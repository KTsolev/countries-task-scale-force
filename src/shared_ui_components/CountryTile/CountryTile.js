
import React, { useRef } from 'react';
import { useDispatch } from 'react-redux';
import { beginUserAction, cancelUserAction } from '../../redux/countriesSlice';

export const CountryTile = (props) => {
    const { country, cssClassName, dissableActions } = props;
    const dispatch = useDispatch();
    const div = useRef(null);
    
    const animateElement = () => {
        if (div.current && !dissableActions) {

            dispatch(beginUserAction(country));
            div.current.className = div.current.className + ' animate';
        }
    }

    const removeAnimation = () => {
        if (div.current && div.current.className && div.current.className.includes('animate')) {

            dispatch(cancelUserAction(null));
            div.current.className = div.current.className.replace('animate', '');
        }
    }

    if (!country) {
        return null;
    }

    return (
        <div 
            className={cssClassName ? `${cssClassName} country-tile`: "country-tile"}
            onMouseDownCapture={animateElement}
            onMouseUpCapture={removeAnimation}>
            <div ref={div} className='country-tile-inner'></div>
            <span className='country-tile-item'>{country.iso2Code || 'N/A'}</span>
            <span className='country-tile-item'>{country.name || 'N/A'}</span>    
            <span className='country-tile-item'>{country.capitalCity || 'N/A'}</span>    
            <span className='country-tile-item'>{country.longitude || 'N/A'}</span>    
            <span className='country-tile-item'>{country.latitude || 'N/A'}</span>
        </div>
    );
}