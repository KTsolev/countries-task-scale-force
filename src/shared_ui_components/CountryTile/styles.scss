
.country {

    &-tile {
        
        @include row;
        position: relative;

        &-inner {
            display: inline-block;
            position: absolute;
            left: 0;
            top: 0;
            z-index: -2;
            width: 0%;
            height: 100%;
            background-color: #80b182;
        }
        
        &:nth-child(2n+1) {
            background-color: #76bbe2;
        }
    
        &-item {
            @include row-item;
            padding: 4px;
            color: #387a75;
        }
    }
}

.animate {
    width: 100%;
    animation: fill;
    animation-duration: 4s;
    animation-timing-function: ease-in-out;
}

@media screen and (max-width: 450px) {
    .country-tile {
       @include column;
    }
    
}

@keyframes fill {
    from {width: 0%};
    to {width: 100%};
}


