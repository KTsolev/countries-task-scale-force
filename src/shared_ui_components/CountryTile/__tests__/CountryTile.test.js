import React from 'react';
import { CountryTile } from '../CountryTile';
import renderer, { act } from 'react-test-renderer';
// eslint-disable-next-line jest/no-mocks-import
import jsonCountries from '../../../__mocks__/mockedParsedJSON';

function createNodeMock(element) {
    if (element.type === 'div') {
      // This is your fake DOM node for <div>.
      // Feel free to add any stub methods, e.g. focus() or any
      // other methods necessary to prevent crashes in your components.
      return {};
    }
    // You can return any object from this method for any type of DOM component.
    // React will use it as a ref instead of a DOM node when snapshot testing.
    return null;
  }

  const mockDispatchFn = jest.fn();

  jest.mock('react-redux', () => ({
    useSelector: selector => selector(),
    useDispatch: () => mockDispatchFn
  }));


describe('CountryTile tests', () => {
  it('renders correctly wtihout props', () => {
    const tree = renderer.create(<CountryTile />);

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly with props', () => {
    const tree = renderer.create(<CountryTile country={jsonCountries.items[2]} cssClassName='row' />);
    
    expect(tree.root.findAllByType('div')[0].props.className).toEqual('row country-tile');

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should handle onmousedown and onmouseup', () => {
    const options = {createNodeMock};
    
    const tree = renderer.create(<CountryTile country={jsonCountries.items[2]} cssClassName='row' />, options);
    
    let button = tree.root.findAllByProps({className: 'row country-tile'})[0];

    const action1 = {
        "payload": {
            "adminregion": {
                "id": "SAS", 
                "iso2code": "8S",
                 "value": "South Asia"
            }, 
            "capitalCity": 
            "Kabul", 
            "id": "AFG", 
            "incomeLevel": {
                "id": "LIC", 
                "iso2code": "XM", 
                "value": "Low income"
            }, 
            "iso2Code": "AF", 
            "latitude": "34.5228", 
            "lendingType": {
                "id": "IDX", 
                "iso2code": "XI", 
                "value": "IDA"
            }, 
            "longitude": "69.1761", 
            "name": "Afghanistan", 
            "region": {
                "id": "SAS", 
                "iso2code": "8S", 
                "value": "South Asia"}
            }, 
            "type": "countries/beginUserAction"};

    const action2 = {"payload": null, "type": "countries/cancelUserAction"}
      
    act(button.props.onMouseDownCapture);

    expect(mockDispatchFn).toHaveBeenCalledWith(action1);
    
    act(button.props.onMouseUpCapture);

    expect(mockDispatchFn).toHaveBeenLastCalledWith(action2);
  });

  
  it('should not handle onmousedown and onmouseup when dissableActions prop is passed', () => {
    const options = {createNodeMock};
    
    const tree = renderer
      .create(<CountryTile country={jsonCountries.items[2]} cssClassName='row' dissableActions />, options);
    
    let button = tree.root.findAllByProps({className: 'row country-tile'})[0];
      
    act(button.props.onMouseDownCapture);

    expect(mockDispatchFn).toHaveBeenCalledTimes(0);

    act(button.props.onMouseUpCapture);

    expect(mockDispatchFn).toHaveBeenCalledTimes(0);
  });
});