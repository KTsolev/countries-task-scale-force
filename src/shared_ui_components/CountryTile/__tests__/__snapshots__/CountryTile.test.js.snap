// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryTile tests renders correctly with props 1`] = `
<div
  className="row country-tile"
  onMouseDownCapture={[Function]}
  onMouseUpCapture={[Function]}
>
  <div
    className="country-tile-inner"
  />
  <span
    className="country-tile-item"
  >
    AF
  </span>
  <span
    className="country-tile-item"
  >
    Afghanistan
  </span>
  <span
    className="country-tile-item"
  >
    Kabul
  </span>
  <span
    className="country-tile-item"
  >
    69.1761
  </span>
  <span
    className="country-tile-item"
  >
    34.5228
  </span>
</div>
`;

exports[`CountryTile tests renders correctly wtihout props 1`] = `null`;
